src/main/java/io/opensergo/OpenSergoClient.java

```diff
@@ -56,17 +59,72 @@ public OpenSergoClient(String host, int port) {
         this.configCache = new SubscribedConfigCache();
         this.subscribeRegistry = new SubscribeRegistry();
         this.reqId = new AtomicInteger(0);
+        status = OpenSergoClientStatus.INITIAL;
+    }
+
+    public void registerSubscribeInfo(OpenSergoClientSubscribeInfo subscribeInfo) {
+        // Register subscriber to local.
+        if (Optional.of(subscribeInfo.getSubscriberList()).isPresent() && subscribeInfo.getSubscriberList().size() > 0) {
+            subscribeInfo.getSubscriberList().forEach(subscriber -> {
+                this.subscribeRegistry.registerSubscriber(subscribeInfo.getSubscribeKey(), subscriber);
+                OpenSergoLogger.info("OpenSergo subscribeinfo registered, subscribeKey={}, subscriber={}", subscribeInfo.getSubscribeKey(), subscriber);
+
+                if (requestAndResponseWriter != null && this.status == OpenSergoClientStatus.STARTED) {
+                    this.subscribeConfig(subscribeInfo.getSubscribeKey());
+                }
+            });
+        }
     }
 
     public void start() throws Exception {
+        OpenSergoLogger.info("OpensergoClient is starting...");
+
+        if (status == OpenSergoClientStatus.INITIAL) {
+            OpenSergoLogger.info("open keepavlive thread");
+            Thread keepAliveThread = new Thread(this::keepAlive);
+            keepAliveThread.setName("thread-opensergo-keepalive-" + keepAliveThread.getId());
+            keepAliveThread.setDaemon(true);
+            keepAliveThread.start();
+        }
+
+        status = OpenSergoClientStatus.STARTING;
+
         this.requestAndResponseWriter = transportGrpcStub.withWaitForReady()
-            .subscribeConfig(new OpenSergoSubscribeClientObserver(configCache, subscribeRegistry));
+            .subscribeConfig(new OpenSergoSubscribeClientObserver(this));
+
+        OpenSergoLogger.info("begin to subscribe config-data...");
+        this.subscribeRegistry.getSubscriberKeysAll().forEach(subscribeKey -> {
+            this.subscribeConfig(subscribeKey);
+        });
+
+        OpenSergoLogger.info("openSergoClient is started");
+        status = OpenSergoClientStatus.STARTED;
+    }
+
+    private void keepAlive() {
+        try {
+            if (status != OpenSergoClientStatus.STARTING
+                    && status != OpenSergoClientStatus.STARTED
+                    && status != OpenSergoClientStatus.SHUTDOWN) {
+                OpenSergoLogger.info("try to restart openSergoClient...");
+                this.start();
+            }
+            Thread.sleep(TimeUnit.SECONDS.toMillis(10));
+            if( status != OpenSergoClientStatus.SHUTDOWN) {
+                keepAlive();
```

Comment:
Maybe we'll need to refactor it with event-based design. Prev discissions: https://github.com/opensergo/opensergo-go-sdk/pull/2#discussion_r1027707654

Reply:
Avoid infinite recursive calls here.

Reply:
Yeah, i will change to for-loop mode to realize the keepalive at current time, and then refactor to event-based design.